### PR TITLE
AP_Proximity: rewrite RPLidarA2 to use more efficient read() method

### PR DESCRIPTION
**Pull request by peterbarker**

### CODE REVIEW

File: libraries/AP_Proximity/AP_Proximity_RPLidarA2.cpp [hunk 1]
```diff
@@ -101,260 +95,228 @@ float AP_Proximity_RPLidarA2::distance_min() const
...
 }
 
-void AP_Proximity_RPLidarA2::get_readings()
+void AP_Proximity_RPLidarA2::consume_bytes(uint16_t count)
 {
-    if (_uart == nullptr) {
-        return;
+    _byte_count -= count;
+    if (_byte_count) {
+        memcpy((void*)&_payload[0], (void*)&_payload[count], _byte_count);
```
Comment (peterbarker) [inline libraries/AP_Proximity/AP_Proximity_RPLidarA2.cpp:129]: `memmove`....

**Comment by amilcarlucas** [comment]
@peterbarker can you add PWM output for RP-Lidar A2 rotation speed control? Maybe in a follow-up PR?
This is nomal PWM, not "RC" PWM

File: libraries/AP_Proximity/AP_Proximity_RPLidarA2.cpp [hunk 2]
```diff
@@ -63,27 +63,21 @@
...
-        get_readings();
-    }
+    get_readings();
 
     // check for timeout and set health status
-    if ((_last_distance_received_ms == 0) || (AP_HAL::millis() - _last_distance_received_ms > COMM_ACTIVITY_TIMEOUT_MS)) {
+    if (AP_HAL::millis() - _last_distance_received_ms > COMM_ACTIVITY_TIMEOUT_MS) {
         set_status(AP_Proximity::Status::NoData);
         Debug(1, "LIDAR NO DATA");
+        if (AP_HAL::millis() - _last_reset_ms > 1000) {
```
Comment (khancyr) [inline libraries/AP_Proximity/AP_Proximity_RPLidarA2.cpp:78]: can't you juste const auto now = AP_HAL::millis() ?

Comment (peterbarker) [reply libraries/AP_Proximity/AP_Proximity_RPLidarA2.cpp]: Done, thanks!

File: libraries/AP_Proximity/AP_Proximity_RPLidarA2.cpp (hunk 1 above)
Comment (peterbarker) [reply libraries/AP_Proximity/AP_Proximity_RPLidarA2.cpp]: Done

**Comment by peterbarker** [comment]
> @peterbarker can you add PWM output for RP-Lidar A2 rotation speed control? Maybe in a follow-up PR?
> This is nomal PWM, not "RC" PWM

@amilcarlucas that sounds interesting.  Is that documented in the usual PDFs?  How would you imagine we use it in AP?

**Comment by peterbarker** [comment]
I've now tested this thoroughly, both in simulation (including a new autotest), and on an instance of the A2 hardware itself.

This is now structured such that it is relatively straight forward to add support for other RPLidar sensors.  I've added support for the A1 based on this PR in this branch: https://github.com/peterbarker/ardupilot/tree/prx/rplidar-a1-support (tested only in autotest!)

I'm hoping we can move on and support the A3 (https://github.com/ArduPilot/ardupilot/pull/10303 @muramura ) and S1 (https://github.com/ArduPilot/ardupilot/pull/14527 @EosBandi ).  Further, this will obsolete #10164 as we'll know the range from the supplied device metadata.

**Comment by amilcarlucas** [comment]
@peterbarker that is the MOTOCTL pin described on page 13 of the [RPLidar A2 datasheet](https://www.digikey.com/htmldatasheets/production/2801237/0/0/1/rplidar-a2-a2m7-a2m8.html)

The PR uses the "High Speed Sampling Protocol and Compatibility" described in page 7 of that datasheet?

File: libraries/AP_Proximity/AP_Proximity_RPLidarA2.cpp [hunk 3]
```diff
@@ -101,260 +107,236 @@ float AP_Proximity_RPLidarA2::distance_min() const
...
+        // this is bad; we were told there were bytes available
+        INTERNAL_ERROR(AP_InternalError::error_t::flow_of_control);
+        reset();
+        return;
+    }
+    _byte_count += bytes_read;
+
+    const uint32_t start_ms = AP_HAL::millis();
+    while (_byte_count) {
+        if (AP_HAL::millis() - start_ms > 300) {
```
Comment (tridge) [inline libraries/AP_Proximity/AP_Proximity_RPLidarA2.cpp:194]: this is bad for flash sector erases

**Review by tridge — CHANGES_REQUESTED** [review]

**Comment by tridge** [comment]
it would be good to add artifical byte loss in testing

File: libraries/AP_Proximity/AP_Proximity_RPLidarA2.cpp [hunk 4]
```diff
@@ -101,260 +107,236 @@ float AP_Proximity_RPLidarA2::distance_min() const
...
+            char xbuffer[64]{};
+            memcpy((void*)xbuffer, (void*)&_payload.information, 63);
+            gcs().send_text(MAV_SEVERITY_INFO, "RPLidar: (%s)", xbuffer);
+#endif
+            // 63 is the magic number of bytes in the spewed-out
+            // reset data ... so now we'll just drop that stuff on
+            // the floor.
+            consume_bytes(63);
+            send_scan_mode_request();
+//            send_request_for_health();
```
Comment (rmackay9) [inline libraries/AP_Proximity/AP_Proximity_RPLidarA2.cpp:225]: I guess we can remove this commented out line..

**Comment by peterbarker** [comment]
As discussed at devcall, changed the too-long-in-loop thing out for a we-must-have-processed-bytes thing, tested it with:

```
         previous_loop_byte_count = _byte_count;
 
+        if (AP_HAL::millis() > 60000) {
+            gcs().send_text(MAV_SEVERITY_INFO, "Breaking it");
+            continue;
+        }
```

and it triggered the internal error.

File: libraries/AP_Proximity/AP_Proximity_RPLidarA2.cpp (hunk 3 above)
Comment (peterbarker) [reply libraries/AP_Proximity/AP_Proximity_RPLidarA2.cpp:194]: Fixed.

File: libraries/AP_Proximity/AP_Proximity_RPLidarA2.cpp (hunk 4 above)
Comment (peterbarker) [reply libraries/AP_Proximity/AP_Proximity_RPLidarA2.cpp:225]: Done

**Comment by peterbarker** [comment]
I've tested this on a real device with simulated byte loss, and the results are good.

**Comment by zenuavsolutions** [comment]
We have massive problems with the A3M1 in sunlight, the device then reports obstacles all around where there are none and the copter starts to fly back and forth too unsteadily. A request to Slamtec showed that the RPLidar has to be set to outdoor mode (stability). Does this patch solve the problem?

**Comment by peterbarker** [comment]
Closing this in favour of https://github.com/ArduPilot/ardupilot/pull/23749/files which incorporates this change

**Pull request closed without merging**